### PR TITLE
feat: replaced entityRecognizers with externalEntityRecognizer in OrchestratorAdaptiveRecognizer

### LINKED ISSUE #3115: port: Align OrchestratorRecognizer with LuisRecognizer ExternalEntityRecognizer (#4995)
The changes in [Align OrchestratorRecognizer with LuisRecognizer ExternalEntityRecognizer (#4995)](https://github.com/microsoft/botbuilder-dotnet/pull/4995) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
## Description
In 4.11 we deprecated the specific EntityRecognizer concept and instead only have Recognizer (EntityRecognizer base class is now Recognizer).  LuisRecognizer has property **ExternalEntityRecognizer** which is of type Recognizer.  This delta aligns with LUIS, so that you can a recognizer which is run and the results passed into orchestrator for orchestrator to use.

## Specific Changes
- Replaced **List<EntityRecognizer> EntityRecognizers** property with **Recognizer ExternalEntityRecognizer**
- updated schema

## Testing
- updated unit tests
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by chon219**
Fixes #3115 

## Description

Replaced entityRecognizers with externalEntityRecognizer in OrchestratorAdaptiveRecognizer

## Specific Changes

  - replaced _entityRecognizers_ with _externalEntityRecognizer_ in `OrchestratorAdaptiveRecognizer`
  - updated `OrchestratorRecognizer`
  - moved _createChooseIntentResult_ to `Recognizer`
  - removed unused dependencies in `botbuilder-ai-orchestrator`

## Testing

Updated test cases for orchestrator

### CODE REVIEW

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 1]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
 
     private readonly unknownIntentFilterScore = 0.4;
     private static orchestrator: any = null;
-    private resolver: any = null;
+    private _resolver: any = null;
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:93]: Would it be simply to fix the `any` types here?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 2]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
-        this._snapshotPath = this.snapshotPath.getValue(dialogContext.state);
-        const detectAmbiguity = this.detectAmbiguousIntents.getValue(dialogContext.state);
-        const disambiguationScoreThreshold = this.disambiguationScoreThreshold.getValue(dialogContext.state);
+        const text = activity.text ?? '';
+        this._modelPath = this.modelPath.getValue(dc.state);
+        this._snapshotPath = this.snapshotPath.getValue(dc.state);
+        const detectAmbiguity = this.detectAmbiguousIntents.getValue(dc.state);
 
-        const recognizerResult: RecognizerResult = createRecognizerResult(text, {}, {});
+        let recognizerResult: RecognizerResult = {
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:143]: Is this var reassigned?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 3]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
-                        closestText: item.closest_text,
-                        result: itemRecoResult,
-                    });
-                });
+            // if top scoring intent is less than threshold, return None
+            if (topScore < this.unknownIntentFilterScore) {
+                recognizerResult.intents.None = { score: 1.0 };
+            } else {
+                // add top score
+                recognizerResult.intents[`${topScoringIntent}`] ??= {
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:182]: Why is the interpolation necessary?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 4]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
+                                (result): RecognizerResult => ({
+                                    text,
+                                    alteredText: result.closest_text,
+                                    entities: recognizerResult.entities,
+                                    intents: { [result.label.name]: { score: result.score } },
+                                })
+                            )
+                            .reduce((results: Record<string, RecognizerResult>, result: RecognizerResult) => {
+                                const guid = generate_guid();
+                                results[`${guid}`] = result;
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:205]: Same question - is the interpolation necessary? `results[guid]` should be fine right?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 5]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
-        // if top scoring intent is less than threshold, return None
-        if (topScore < this.unknownIntentFilterScore) {
-            recognizerResult.intents['None'] = { score: 1.0 };
-        } else if (!recognizerResult.intents[topScoringIntent]) {
-            recognizerResult.intents[topScoringIntent] = {
-                score: topScore,
-            };
+    private _initializeModel() {
+        if (!this._modelPath) {
+            throw new Error(`Missing "ModelPath" information.`);
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:237]: Is there any reason not to perform this validation in the constructor?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 6]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
-                startIndex: entityResult['start'],
-                endIndex: entityResult['end'],
-                score: 1.0,
-                text: entityResult['text'],
-                type: entityResult['type'],
-                resolution: entityResult['resolution'],
-            };
-            instanceData.push(instance);
+            OrchestratorAdaptiveRecognizer.orchestrator = orchestrator;
+            console.timeEnd('Model load');
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:256]: These timings should be removed before merging, right?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 7]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
+        if (!this._resolver) {
             const fullSnapshotPath = resolve(this._snapshotPath);
-            if (!existsSync(fullModelPath)) {
-                throw new Error(`Model folder does not exist at ${fullModelPath}.`);
-            }
             if (!existsSync(fullSnapshotPath)) {
                 throw new Error(`Snapshot file does not exist at ${fullSnapshotPath}.`);
             }
+            // Load the snapshot
+            const snapshot: Uint8Array = readFileSync(fullSnapshotPath) as Buffer;
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:265]: Why the casting?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts [hunk 8]
```diff
@@ -110,228 +90,196 @@ export class OrchestratorAdaptiveRecognizer extends Recognizer implements Orches
...
         }
     }
 }
+
+/*
+ * This function generates a GUID-like random number that should be sufficient for our purposes of tracking
+ * instances of a given waterfall dialog.
+ * Source: https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+ */
+function generate_guid() {
```
Comment (joshgummersall) [inline libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:278]: Let's get this function added to `botbuilder-stdlib` I see it all over the place.

**Review by joshgummersall — COMMENTED** [review]
Couple quick comments.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 3 above)
Comment (chon219) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:182]: I got a lint warning here: 
image

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 5 above)
Comment (chon219) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:237]: In a declarative based bot, the resource explorer always construct components with empty parameters, so the validation cannot be done in constructor.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 7 above)
Comment (chon219) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:265]: Casting is not required, removing this.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 8 above)
Comment (chon219) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:278]: Yes, shall we send out another PR for this?

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 2 above)
Comment (chon219) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:159]: Yes, it is reassigned in line 226.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 5 above)
Comment (joshgummersall) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:253]: Makes sense!

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 8 above)
Comment (joshgummersall) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:292]: Sure, sounds good.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 3 above)
Comment (joshgummersall) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:198]: Ah - we really should tune or disable that lint rule. I believe it's intended to protect against using unsanitized user input for object keys, which isn't that concerning. Feel free to add `// eslint-disable-next-line security/detect-object-injection` to remove the lint warning.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 8 above)
Comment (joshgummersall) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:292]: Added https://github.com/microsoft/botbuilder-js/issues/3160 to track.

File: libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts (hunk 3 above)
Comment (joshgummersall) [reply libraries/botbuilder-ai-orchestrator/src/orchestratorAdaptiveRecognizer.ts:198]: Submitted https://github.com/microsoft/botbuilder-js/issues/3161 to track.

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**